File: jm/diag/canbus.cpp
#include "canbus.h"
#include "../utils.h"

namespace JM
{
namespace Diag
{

Canbus::Canbus()
{
    _flowControl[0] = 0x30;
    _flowControl[1] = 0x00;
    _flowControl[2] = 0x00;
    _flowControl[3] = 0x00;
    _flowControl[4] = 0x00;
    _flowControl[5] = 0x00;
    _flowControl[6] = 0x00;
    _flowControl[7] = 0x00;
}

Canbus::~Canbus()
{

}

boost::system::error_code Canbus::config(const Canbus::Options &opts, boost::system::error_code &ec)
{
    _opts.baud = opts.baud;
    _opts.filterMask = opts.filterMask;
    _opts.frameType = opts.frameType;
    _opts.highPin = opts.highPin;
    _opts.idMode = opts.idMode;
    _opts.idVector = opts.idVector;
    _opts.lowPin = opts.lowPin;
    _opts.targetID = opts.targetID;
    ec = boost::system::error_code();
    return ec;
}

std::size_t Canbus::pack(const boost::uint8_t *src,
                         std::size_t srcLength,
                         boost::uint8_t *tar,
                         std::size_t tarMaxLength,
                         boost::system::error_code &ec)
{
    if (srcLength > 8 || srcLength <= 0)
    {
        ec = boost::asio::error::message_size;
        return 0;
    }

    if (_opts.idMode == StdID)
    {
        tar[1] = high_byte(low_word(_opts.targetID));
        tar[2] = low_byte(low_word(_opts.targetID));
        if (_opts.frameType == DataFrame)
        {
            tar[0] = low_byte(srcLength | StdID | DataFrame);
        }
        else
        {
            tar[0] = low_byte(srcLength | StdID | RemoteFrame);
        }
        memcpy(tar + 3, src, srcLength);
        return srcLength + 3;
    }
    if (_opts.idMode == ExtID)
    {
        tar[1] = high_byte(high_word(_opts.targetID));
        tar[2] = low_byte(high_word(_opts.targetID));
        tar[3] = high_byte(low_word(_opts.targetID));
        tar[4] = low_byte(low_word(_opts.targetID));
        if (_opts.frameType == DataFrame)
        {
            tar[0] = low_byte(srcLength | ExtID | DataFrame);
        }
        else
        {
            tar[0] = low_byte(srcLength | ExtID | RemoteFrame);
        }
        memcpy(tar + 5, src, srcLength);
        return srcLength + 5;
    }
    return 0;
}

std::size_t Canbus::unpack(const boost::uint8_t *src,
                           std::size_t srcLength,
                           boost::uint8_t *tar,
                           std::size_t tarMaxLength,
                           boost::system::error_code &ec)
{
    if (srcLength <= 0)
    {
        ec = boost::asio::error::message_size;
        return 0;
    }

    std::size_t length = 0;

    boost::uint32_t mode = (src[0] & (ExtID | RemoteFrame));
    if (mode == (StdID | DataFrame))
    {
        length = src[0] & 0x0F;
        if (length != srcLength - 3)
        {
            ec = boost::asio::error::message_size;
            return 0;
        }

        memcpy(tar, src + 3, length);
    }

    if (mode == (ExtID | DataFrame))
    {
        length = src[0] & 0x0F;
        if (length != srcLength - 5)
        {
            ec = boost::asio::error::message_size;
            return 0;
        }

        memcpy(tar, src + 5, length);
    }
    return length;
}

}
}

File: JM/Diag/Canbus.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using JM.Core;

namespace JM.Diag
{
    public class Canbus : IProtocol
    {
        private byte[] flowControl = new byte[8];
        protected CanbusOptions options;

        public Canbus()
        {
            flowControl[0] = 0x30;
            flowControl[1] = 0x00;
            flowControl[2] = 0x00;
            flowControl[3] = 0x00;
            flowControl[4] = 0x00;
            flowControl[5] = 0x00;
            flowControl[6] = 0x00;
            flowControl[7] = 0x00;
        }

        public byte[] Pack(byte[] data, int offset, int count)
        {
            if (count > 8 || count <= 0)
            {
                return null;
            }

            byte[] result = null;

            if (options.IDMode == CanbusIDMode.Standard)
            {
                result = new byte[3 + count];
                result[1] = Utils.HighByte(Utils.LowWord(options.TargetID));
                result[2] = Utils.LowByte(Utils.LowWord(options.TargetID));
                if (options.FrameType == CanbusFrameType.Data)
                {
                    result[0] = Utils.LowByte(count | (int)CanbusIDMode.Standard | (int)CanbusFrameType.Data);
                }
                else
                {
                    result[0] = Utils.LowByte(count | (int)CanbusIDMode.Standard | (int)CanbusFrameType.Remote);
                }
                Array.Copy(data, offset, result, 3, count);
            }
            else
            {
                result = new byte[5 + count];
                result[1] = Utils.HighByte(Utils.HighWord(options.TargetID));
                result[2] = Utils.LowByte(Utils.HighWord(options.TargetID));
                result[3] = Utils.HighByte(Utils.LowWord(options.TargetID));
                result[4] = Utils.LowByte(Utils.LowWord(options.TargetID));
                if (options.FrameType == CanbusFrameType.Data)
                {
                    result[0] = Utils.LowByte(count | (int)CanbusIDMode.Extension | (int)CanbusFrameType.Data);
                }
                else
                {
                    result[0] = Utils.LowByte(count | (int)CanbusIDMode.Extension | (int)CanbusFrameType.Remote);
                }
                Array.Copy(data, offset, result, 5, count);
            }

            return result;
        }

        public byte[] Unpack(byte[] data, int offset, int count)
        {
            if (count <= 0)
            {
                return null;
            }

            int length = 0;
            int mode = (data[offset] & ((int)CanbusIDMode.Extension | (int)CanbusFrameType.Remote));
            byte[] result = null;
            if (mode == ((int)CanbusIDMode.Standard | (int)CanbusFrameType.Data))
            {
                length = data[offset] & 0x0F;
                if (length != count - 3)
                {
                    return null;
                }

                result = new byte[length];
                Array.Copy(data, offset, result, 0, length);
            }
            else if(mode == ((int)CanbusIDMode.Extension | (int)CanbusFrameType.Remote))
            {
                length = data[offset] & 0x0F;
                if (length != count - 5)
                {
                    return null;
                }
                result = new byte[length];
                Array.Copy(data, offset, result, 0, length);
            }

            return result;
        }

        public virtual void Config(ref CanbusOptions opts)
        {
            options.Baud = opts.Baud;
            options.FilterMask = opts.FilterMask;
            options.FrameType = opts.FrameType;
            options.HighPin = opts.HighPin;
            options.IDMode = opts.IDMode;
            options.IDVector = opts.IDVector;
            options.LowPin = opts.LowPin;
            options.TargetID = opts.TargetID;
        }
    }
}
